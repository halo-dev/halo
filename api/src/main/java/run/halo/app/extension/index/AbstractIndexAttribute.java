package run.halo.app.extension.index;

import lombok.EqualsAndHashCode;
import org.springframework.util.Assert;
import run.halo.app.extension.Extension;
import run.halo.app.extension.GVK;

@EqualsAndHashCode
public abstract class AbstractIndexAttribute<E extends Extension> implements IndexAttribute {
    private final Class<E> objectType;

    /**
     * Creates a new {@link AbstractIndexAttribute} for the given object type.
     *
     * @param objectType must not be {@literal null}.
     */
    public AbstractIndexAttribute(Class<E> objectType) {
        Assert.notNull(objectType, "Object type must not be null");
        Assert.state(isValidExtension(objectType),
            "Invalid extension type, make sure you have annotated it with @" + GVK.class
                .getSimpleName());
        this.objectType = objectType;
    }

    @Override
    public Class<E> getObjectType() {
        return this.objectType;
    }

    boolean isValidExtension(Class<? extends Extension> type) {
        return type.getAnnotation(GVK.class) != null;
    }
}
