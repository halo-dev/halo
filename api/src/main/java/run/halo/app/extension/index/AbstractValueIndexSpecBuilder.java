package run.halo.app.extension.index;

import org.springframework.util.Assert;
import run.halo.app.extension.Extension;

/**
 * Abstract base implementation of {@link IndexSpecBuilder} for value indexes.
 *
 * @param <E> the type of extension
 * @param <K> the type of key
 * @param <B> the type of builder
 * @author johnniang
 * @since 2.22.0
 */
abstract class AbstractValueIndexSpecBuilder<
    E extends Extension,
    K extends Comparable<K>,
    B extends IndexSpecBuilder<E, K, B>
    > implements IndexSpecBuilder<E, K, B> {

    protected final String name;

    protected final Class<K> keyType;

    protected boolean unique = false;

    protected boolean nullable = true;

    protected AbstractValueIndexSpecBuilder(String name, Class<K> keyType) {
        Assert.hasText(name, "Index name must not be blank");
        Assert.notNull(keyType, "Key type must not be null");
        this.name = name;
        this.keyType = keyType;
    }

    public B unique(boolean unique) {
        this.unique = unique;
        return (B) this;
    }

    public B nullable(boolean nullable) {
        this.nullable = nullable;
        return (B) this;
    }

}
