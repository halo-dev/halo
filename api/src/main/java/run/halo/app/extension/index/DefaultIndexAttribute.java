package run.halo.app.extension.index;

import java.util.Optional;
import java.util.Set;
import java.util.function.Function;
import org.springframework.util.Assert;
import run.halo.app.extension.Extension;

@Deprecated(forRemoval = true, since = "2.22.0")
class DefaultIndexAttribute<E extends Extension, K extends Comparable<K>>
    implements IndexAttribute<E, K> {

    private final Class<E> objectType;

    private final Class<K> keyType;

    private final Function<E, Set<K>> valuesFunc;

    public DefaultIndexAttribute(Function<E, Set<K>> valuesFunc, Class<E> objectType,
        Class<K> keyType) {
        Assert.notNull(valuesFunc, "Values function must not be null");
        Assert.notNull(objectType, "Cannot resolve object type");
        Assert.notNull(keyType, "Cannot resolve key type");
        this.valuesFunc = valuesFunc;
        this.objectType = objectType;
        this.keyType = keyType;
    }

    @Override
    public Class<E> getObjectType() {
        return objectType;
    }

    @Override
    public Class<K> getKeyType() {
        return keyType;
    }

    @Override
    public Set<K> getValues(E e) {
        if (!checkType(e)) {
            throw new IllegalArgumentException("Object type does not match");
        }
        return Optional.ofNullable(this.valuesFunc.apply(e)).orElse(Set.of());
    }

    private boolean checkType(Extension object) {
        return getObjectType().isInstance(object);
    }
}
