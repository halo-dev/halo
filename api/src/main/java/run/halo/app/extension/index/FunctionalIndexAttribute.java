package run.halo.app.extension.index;

import java.util.Set;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import org.springframework.lang.Nullable;
import org.springframework.util.Assert;
import run.halo.app.extension.Extension;

@EqualsAndHashCode(callSuper = true)
public class FunctionalIndexAttribute<E extends Extension>
    extends AbstractIndexAttribute<E> {

    @EqualsAndHashCode.Exclude
    private final Function<E, String> valueFunc;

    /**
     * Creates a new {@link FunctionalIndexAttribute} for the given object type and value function.
     *
     * @param objectType must not be {@literal null}.
     * @param valueFunc value function must not be {@literal null}.
     */
    public FunctionalIndexAttribute(Class<E> objectType,
        Function<E, String> valueFunc) {
        super(objectType);
        Assert.notNull(valueFunc, "Value function must not be null");
        this.valueFunc = valueFunc;
    }

    @Override
    public Set<String> getValues(Extension object) {
        var value = getValue(object);
        return value == null ? Set.of() : Set.of(value);
    }

    /**
     * Gets the value for the given object.
     *
     * @param object the object to get the value for.
     * @return returns the value for the given object.
     */
    @Nullable
    public String getValue(Extension object) {
        if (getObjectType().isInstance(object)) {
            return valueFunc.apply(getObjectType().cast(object));
        }
        throw new IllegalArgumentException("Object type does not match");
    }
}
