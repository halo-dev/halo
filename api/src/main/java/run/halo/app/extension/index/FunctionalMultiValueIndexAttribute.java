package run.halo.app.extension.index;

import java.util.Set;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import org.springframework.util.Assert;
import run.halo.app.extension.Extension;
import run.halo.app.extension.Unstructured;

@EqualsAndHashCode(callSuper = true)
public class FunctionalMultiValueIndexAttribute<E extends Extension>
    extends AbstractIndexAttribute<E> {

    @EqualsAndHashCode.Exclude
    private final Function<E, Set<String>> valueFunc;

    /**
     * Creates a new {@link FunctionalIndexAttribute} for the given object type and value function.
     *
     * @param objectType object type must not be {@literal null}.
     * @param valueFunc value function must not be {@literal null}.
     */
    public FunctionalMultiValueIndexAttribute(Class<E> objectType,
        Function<E, Set<String>> valueFunc) {
        super(objectType);
        Assert.notNull(valueFunc, "Value function must not be null");
        this.valueFunc = valueFunc;
    }

    @Override
    public Set<String> getValues(Extension object) {
        if (object instanceof Unstructured unstructured) {
            var ext = Unstructured.OBJECT_MAPPER.convertValue(unstructured, getObjectType());
            return getNonNullValues(ext);
        }
        if (getObjectType().isInstance(object)) {
            return getNonNullValues(getObjectType().cast(object));
        }
        throw new IllegalArgumentException("Object type does not match");
    }

    private Set<String> getNonNullValues(E object) {
        var values = valueFunc.apply(object);
        return values == null ? Set.of() : values;
    }
}
