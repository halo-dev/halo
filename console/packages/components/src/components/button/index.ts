export { default as VButton } from "./Button.vue";
