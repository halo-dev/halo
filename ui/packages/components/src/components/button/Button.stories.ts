import type { Meta, StoryObj } from "@storybook/vue3";

import { VButton } from ".";

const meta: Meta<typeof VButton> = {
  title: "Button",
  component: VButton,
  tags: ["autodocs"],
  render: (args) => ({
    components: { VButton },
    setup() {
      return { args };
    },
    template: `<VButton v-bind="args">${args.default}</VButton>`,
  }),
  argTypes: {
    default: {
      control: { type: "text" },
    },
    type: {
      control: { type: "select" },
      options: ["default", "primary", "secondary", "danger"],
    },
    size: {
      control: {
        type: "select",
      },
      options: ["lg", "md", "sm", "xs"],
    },
  },
  args: {
    default: "Button",
  },
};

export default meta;
type Story = StoryObj<typeof VButton>;

export const Default: Story = {
  args: {
    type: "default",
  },
};

export const Primary: Story = {
  args: {
    type: "primary",
  },
};

export const Secondary: Story = {
  args: {
    type: "secondary",
  },
};

export const Danger: Story = {
  args: {
    type: "danger",
  },
};
