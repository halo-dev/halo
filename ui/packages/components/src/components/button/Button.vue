<template>
  <button
    :class="classes"
    :disabled="disabled"
    class="btn"
    type="button"
    @click="handleClick"
  >
    <span v-if="$slots.icon || loading" class="btn-icon">
      <svg
        v-if="loading"
        class="animate-spin"
        fill="none"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          class="opacity-25"
          cx="12"
          cy="12"
          r="10"
          stroke="currentColor"
          stroke-width="4"
        ></circle>
        <path
          class="opacity-75"
          d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
          fill="currentColor"
        ></path>
      </svg>
      <slot v-else name="icon" />
    </span>
    <span class="btn-content">
      <slot />
    </span>
  </button>
</template>
<script lang="ts" setup>
import { computed } from "vue";
import type { RouteLocationRaw } from "vue-router";
import { useRouter } from "vue-router";
import type { Size, Type } from "./interface";

const props = withDefaults(
  defineProps<{
    type?: Type;
    size?: Size;
    circle?: boolean;
    block?: boolean;
    disabled?: boolean;
    loading?: boolean;
    route?: RouteLocationRaw | undefined;
  }>(),
  {
    type: "default",
    size: "md",
    circle: false,
    block: false,
    disabled: false,
    loading: false,
    route: undefined,
  }
);

const router = useRouter();
const emit = defineEmits<{
  (event: "click"): void;
}>();

const classes = computed(() => {
  return [
    `btn-${props.size}`,
    `btn-${props.type}`,
    { "btn-circle": props.circle },
    { "btn-block": props.block },
    { "btn-loading": props.loading },
  ];
});

function handleClick() {
  if (props.disabled || props.loading) return;
  if (props.route) {
    router.push(props.route);
  }
  emit("click");
}
</script>
<style lang="scss">
.btn {
  @apply rounded-base
  inline-flex
  flex-shrink-0
  cursor-pointer
  select-none
  flex-wrap
  items-center
  justify-center
  transition-all
  text-center
  text-sm
  no-underline
  h-9
  px-4
  outline-0
  border-none
  appearance-none
  align-middle;

  &:hover {
    @apply opacity-90;
  }

  &:active {
    @apply opacity-100;
  }

  &:disabled {
    @apply opacity-50
    cursor-not-allowed;
  }
}

.btn-default {
  border: 1px solid #d9d9d9;

  &:hover {
    @apply bg-gray-100;
  }

  .btn-icon {
    @apply text-secondary;
  }
}

.btn-primary {
  @apply text-white bg-primary #{!important};
}

.btn-secondary {
  @apply text-white bg-secondary #{!important};
}

.btn-danger {
  background-color: #d71d1d !important;
  @apply text-white;
}

.btn-block {
  @apply w-full;
}

.btn-icon {
  @apply h-5 w-5
  text-white
  mr-3;
}

.btn-loading {
  @apply cursor-not-allowed;
  &:hover {
    @apply opacity-100;
  }
}

.btn-lg {
  @apply h-11
  px-5
  text-lg;
}

.btn-sm {
  @apply h-7
  px-3
  text-xs;

  .btn-icon {
    @apply h-3
    w-3
    mr-2;
  }
}

.btn-xs {
  @apply h-6
  px-2
  text-xs;

  .btn-icon {
    @apply h-3
    w-3
    mr-2;
  }
}

.btn-circle {
  @apply w-9
  p-0
  rounded-full;
}

.btn-lg.btn-circle {
  @apply w-11;
}

.btn-sm.btn-circle {
  @apply w-7;
}

.btn-xs.btn-circle {
  @apply w-6;
}
</style>
