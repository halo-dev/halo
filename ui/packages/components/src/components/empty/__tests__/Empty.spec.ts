import { describe, expect, it } from "vitest";
import { mount } from "@vue/test-utils";
import { VEmpty } from "../index";
import { VButton } from "../../button";
import { h } from "vue";

describe("Empty", () => {
  it("should render", () => {
    expect(mount(VEmpty)).toBeDefined();
  });

  it("should match snapshot", () => {
    expect(
      mount(VEmpty, {
        props: {
          title: "Not found",
          message: "No posts found",
          image: "./Empty.svg",
        },
        slots: {
          actions: h(VButton, { type: "primary" }, "New Post"),
        },
      }).html()
    ).toMatchSnapshot();
  });

  it("should work with title prop", () => {
    const wrapper = mount(VEmpty, { props: { title: "Not found" } });
    expect(wrapper.find(".empty-title").text()).toEqual("Not found");
  });

  it("should work with message prop", () => {
    const wrapper = mount(VEmpty, { props: { message: "No posts found" } });
    expect(wrapper.find(".empty-message").text()).toEqual("No posts found");
  });

  it("should work with message slot", () => {
    const wrapper = mount(VEmpty, {
      props: { message: "empty" },
      slots: { message: h("span", h("storage", "No posts found")) },
    });

    expect(wrapper.find(".empty-message")).not.toEqual("Empty");
    expect(wrapper.find(".empty-message > span > storage").text()).toEqual(
      "No posts found"
    );
  });

  it("should work with actions slot", () => {
    const wrapper = mount(VEmpty, {
      slots: {
        actions: h(VButton, { type: "primary" }, "New Post"),
      },
    });

    expect(wrapper.findComponent(VButton)).toBeDefined();
    expect(wrapper.findComponent(VButton).find(".btn-content").text()).toEqual(
      "New Post"
    );
  });

  it("should work with image prop", async () => {
    const wrapper = mount({
      data() {
        return {
          image: "",
        };
      },
      render() {
        return h(VEmpty, {
          image: this.image,
        });
      },
    });

    expect(
      wrapper.find(".empty-image > img").attributes().src.endsWith("/Empty.svg")
    ).toBe(true);

    await wrapper.setData({ image: "./empty.png" });

    expect(wrapper.find(".empty-image > img").attributes().src).toEqual(
      "./empty.png"
    );
  });

  it("should work with image slot", () => {
    const wrapper = mount(VEmpty, {
      slots: { image: h("img", { src: "./empty", alt: "Empty Status" }) },
    });

    const attributes = wrapper.find(".empty-image > img").attributes();

    expect(attributes.src).not.toEqual("/src/components/empty/Empty.svg");

    expect(attributes.src).toEqual("./empty");
    expect(attributes.alt).toEqual("Empty Status");
  });
});
