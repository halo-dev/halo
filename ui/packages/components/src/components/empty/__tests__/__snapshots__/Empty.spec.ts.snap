// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Empty > should match snapshot 1`] = `
"<div class=\\"empty-wrapper\\">
  <div class=\\"empty-image h-32 w-32\\"><img src=\\"./Empty.svg\\" alt=\\"Empty\\"></div>
  <div class=\\"empty-title\\">Not found</div>
  <div class=\\"empty-message\\">No posts found</div>
  <div class=\\"empty-actions\\"><button class=\\"btn-md btn-primary btn\\" type=\\"button\\">
      <!--v-if--><span class=\\"btn-content\\">New Post</span>
    </button></div>
</div>"
`;
